src/hotspot/share/gc/shenandoah/shenandoahBufferWithSATBRememberedSet.hpp

```diff
@@ -0,0 +1,103 @@
+//         just recently written by this same thread will be available in the L1 cache and fetching this single reference
+//         "now" is more efficient than reading the entire card's memory at remembered set scanning time only to discover
+//         then that the card represents no references to young-gen memory.
+//
+//     Experiments with each approach may help decide between approaches.
+//
+//  3. Potentially, the incremental overhead of adding remembered set logging to the existing SATB barrier is lower than
+//     the incremental overhead of adding an independent and distinct new write barrier for card marking.  This is especially
+//     true during times that the SATB barrier is already enabled (which might represent 30-50% of execution for GC-intensive
+//     workloads).  It may even be true during times that the SATB is disabled.  This is because even when the SATB is disabled,
+//     register allocation is constrained by the reality that SATB will be enabled some of the time, so registers allocated for
+//     SATB-buffer maintenance will sit idle during times when the SATB barrier is disabled.  In tight loops that write pointer
+//     values, for example, the SATB implementation might dedicate registers to holding thread-local information associated with
+//     maintenance of the SATB buffer such as the address of the SATB buffer and the next available buffer slot within this buffer.
+//     In the traditional card-marking remembered set implementation, an additional register might be dedicated to holding a
+//     reference to the base of the card table.
+//
+// Note that a ShenandoahBufferWithSATBRememberedSet maintains the remembered set only for memory regions that correspond to
+// old-generation memory.  Anticipate that the implementation will use double indirection.
+//
+// To Do:
+//
+//  1. Figure out which region an old-gen address belongs to
+//  2. Find the cluster and card numbers that corresponds to that region
+//
+// Old-gen memory is not necessarily contiguous.  It may be comprised of multiple heap regions.
+//
+// The memory overhead for crossing map and card-table entries is given by the following analysis:
+//   For each 512-byte card-entry span, we have the following overhead:
+//    2 bytes for the object_starts map
+//    1 bit for the card-table entry
+//  =====
+//    2 bytes (rounded down) out of 512 bytes is 0.39% bookkeeping overhead
+
+
+// ShenandoahBufferWithSATBRememberedSet is not implemented correctly in its current form.
+//
+// This class is a placeholder to support a possible future implementation of remembered set that uses a generalization of the
+// existing SATB pre-write barrier to maintain remembered set information rather than using unconditional direct card marking in
+// a post-write barrier.
+class ShenandoahBufferWithSATBRememberedSet: public CHeapObj<mtGC> {
```

Comment:
It's also lacking the usual #ifndef - #define - #endif sequence that is needed by headers.

Reply:
This file has been removed from the pull request.